### ISSUE TITLE
CM19a issue

### ISSUE THREAD

**Issue by alzzy** [genielabs/HomeGenie#353]
I am using cm19a in HG on the Pix10hub 1.1 release. Running on Pi Zero W. 

I have my cm15a unplugged (also its RF interface is disabled). I have shutdown my Alex10 app and its HA bridge. I have no other x10 software running and the cm19a is my only x10 interface. I use at v572a tranceiver. I have cm19a selected as the x10 interface in HG. Interface function is erratic. If a command is not sent to cm19a for several seconds, it takes two clicks of an x10 module On or Off button in HG for the cm19a to transmit the command. Its like the first click wakes it up and the second transmits the command. I have tried 2 different cm19a's with the same results.
Note: my RF connection and v572 transceiver response is very reliable as I have very excellent results with cm19a controlling my devices when running it with Alex10. The other thing I noticed is that it appears the flash of the led on the CM19a when sending a command is shorter in duration when using HG than when using Alex10 if that means anything.

Any help is appreciated!

**Comment by mralapete** [comment]
CM19 driver working in HG v1.2-Stable .28 perfectly both on the Raspberry Pi Zero W and the Raspberry Pi 3+ under  Raspbian Stretch Lite. Also working perfectly in Ubuntu 18.04. All commands issued are acted upon instantaneously.

Suggest you look at the additional software you are running in parallel with HG to isolate you problem. You may need to run that additional software on a separate Pi server as the additional overhead on the Pi Zero W may be causing a bottleneck. Alternatively a change to the Raspberry Pi 3+ may be the answer.

You could of course run one of the many Linux resource monitors available to track your Pi performance.

**Comment by tuicemen** [comment]
My Cm19a led doesn't even light up the slightest using HG I actually Had thought the LED had burnt out. I use the cm15 but have tested the cm19.
I too have experienced the need for a double push of the on or off buttons in HG occasionally if the system has been idle for a while. I'll do a bit more testing to see if it is a particular module which may indicate a noise or signal sucker at play here. I'm not sure as yet if this affects macros or timers here.

**Comment by alzzy** [comment]
Tuicemen,
How could it be a module problem if the CM19a doesn't light up? The cm19 isn't transmitting a command, Doesn't it have to be a Pi or HG issue or HG driver issue?

**Comment by tuicemen** [comment]
My Cm19 (unit)LED light has never lite up using HG (that I noticed) but the commands do go threw and modules do come on/off.
RF signals can be affected by a number of things in fact nothing is received if my CM19 is in the basement.
@mralapete I tend to agree however I never ran the CM19 for an extended time with a large configuration. I believe location of the cm19 to be the major issue at play here.
@alzzy is the CM19 in the same location as it was running the other software? If not try placing it there.

**Comment by alzzy** [comment]
My cm19 is in the same position it was in with Alex10. Its not a position or RF issue. I have rock solid RF in my set up. In my set up when the cm19 led does not light up I get no module response , when it does I get module response. Every time. Its either a driver issue or the Pi zero has an issue.

**Comment by mralapete** [comment]
The Pi Zero W is a relatively low speced board to run two separate servers on. The overhead placed on it from additional traffic could well be causing your problem. The Pi 3+ would be a more suitable unit to use here or as I suggested split the load between two separate Pi servers and expose one server to the other via the many methods explained in the HG documentation.

Once again the CM19 is functioning normally.

**Comment by alzzy** [comment]
Well that's a bummer! Tuicemen do you agree?

**Comment by alzzy** [comment]
The little cm19a is overburdening the Pi Zero?? But the cm15a doesn't?

**Comment by tuicemen** [comment]
The cm15 is self powered the cm19 draws power from the Pi
The Zero may not be supplying enough power to give a good RF signal.
As I stated before I've not done enough testing with the CM19 and the Pi Zero I only know HG does send the RF signal to the CM15.
I only have Pi zero W boards here in the city.

**Comment by genemars** [comment]
When the RF signal is weak, sending the same message more than just once can improve the result (maybe the initial RF start burst is not recognized very well).
Even X10 sensors send 4 times the same message (with a very short pause between each), to give the message more chances to be received correctly.
So I can make CM19Lib send the RF message **twice** by default and make this parameter configurable somehow.
Any thoughts?

**Comment by alzzy** [comment]
genemars: Sounds like the thing to do! But I'd vote for **Thrice** just to be sure.

**Comment by alzzy** [comment]
genemars: is sending a command and triggering a cm19a a significant increase in load on the pi processor over doing the same with a cm15a?

**Comment by tuicemen** [comment]
@alzzy I'm not sure how close your pi/cm19 combo is to your  v572a but why not put them next to each other as a test.
I know if I put my cm19 next to a transceiver it works every time the further away the more unreliable it is.
 Also I've now tested the CM19 with 3 different Pi Zero W boards. Some actually light up the LED on the CM19 with a command send. So they definitely don't all send the same power to the CM19 which may explain why some users have such a reliable setup.
One of my Pi ZeroW boards requires a double mouse click of any button in HG, others work with one mouse click this is with the CM19 never changing location.
All boards were using the same Power supply so in my case one ZeroW boards quality is not up to pare.

**Comment by alzzy** [comment]
Maybe the PI Z usb port can't reliably supply power to the cm19 and the thing to try is to provide power to the cm19  directly from a power adapter rather than have the Pi Z supply power thru the usb port. Thoughts? Do you know if anyone has had success with a powered usb hub on a Pi Z Pix10hub?

**Comment by tuicemen** [comment]
A powered USB hub would most likely help although some users have fried their zero W boards with a powered hub. For this reason I won't recommend you try it. If you do attempt this you have been warned should you fry the board.

**Comment by mralapete** [comment]
@genemars what work have you based your CM19 driver on. There's one recent user of HG that is experiencing difficulty using your CM19 driver and there's a third party server in the form of HA Bridge involved in their setup. Maybe if you installed the HA Bridge yourself and tested your driver you could determine if the problem lay with your CM19 driver and not with how comms are handled between the HA Bridge and HG before you start making changes to your code. 

Personally I'm experiencing no issues with the CM19 driver. From what I can gather the OP is trying to establish an Alexa setup here on a Raspberry Pi Zero W using the 3rd party HA Bridge and HG. If its such a popular request maybe you could do away with the need for the 3rd party HA Bridge and integrate Alexa support directly in HG as was being worked on by a number of the more experienced HG users/contributors before development halted. This would give greater appeal to the more popular technologies like ZWave etc

**Comment by alzzy** [comment]
I am confused by the above. Tuicemen, what is your sense of the Zero W issue. USB power to power the cm19a or processing power to command it? If is usb power then perhaps an external supply to power the cm19a could fix it.

**Comment by mralapete** [comment]
Ok. I'll leave it to the X10 North American v European debate so. I don't believe that this is an issue though. What I will say is that any additional overhead on a lower speced board like the Pi Zero W v the Pi 3B+ will introduce performance issues. Fact. I've tried the HA Bridge on the same Pi Zero board as HG and on  a separate Pi Zero board and there is a minor difference. 

Once the correct power supply is used on the Pi Zero W power output to the CM19 will not be an issue so that is not your problem.

As you can imagine the  X10 user base on HG would be quite small so I wouldn't hold out on too many other replies on this issue. Maybe you guys could work this out between the two of you.

**Comment by genemars** [comment]
> @genemars [...] maybe you could do away with the need for the 3rd party HA Bridge and integrate Alexa support directly in HG as was being worked on by a number of the more experienced HG users/contributors before development halted. This would give greater appeal to the more popular technologies like ZWave etc

Yes I might add support for Alexa and other voice controlled gateways directly in HG. But I'm currently working on something else, so not sure when I'll start looking at this.

**Comment by genemars** [comment]
@tuicemen @alzzy even though probably this is not an issue I wanted to make a small modification to CM19Lib and XTenLib to make RF comms more reliable.
I'm publishing the new release shortly.

**Comment by tuicemen** [comment]
@alzzy In my opinion I wouldn't and don't intend on using a CM19 with my Zero W board.
If you want more reliable performance use a CM15.
I plan to test a CM19 with my pi 3B+ which is at my off grid place but that will not be for several months.
@genemars  I'm sure the small modification will be most welcome.
while playing with the CM19 I noticed if I had HG set to use the CM15 it would still see RF the Cm19 received so I wonder if the two (CM19Lib and XTenLib) could some how be combined.

**Comment by genemars** [comment]
> I wonder if the two (CM19Lib and XTenLib) could some how be combined.

Yes it can be done but I don't see any possible use.

**Comment by tuicemen** [comment]
> Yes it can be done but I don't see any possible use.

Other them a smaller download I don't see any benefit except maybe adding a option for a cm15 to send either RF or PLC with a simple switch option in the widget instead of having to create a macro.

However this was just me thinking out loud ;)

**Comment by alzzy** [comment]
I never had reliable rf coms with the cm15. If it were me I would disable the cm15 RF capability and install a V572A transciever. did it 20 years ago and have had great rf coms ever since.

**Comment by tuicemen** [comment]
> I never had reliable rf coms with the cm15. If it were me I would disable the cm15 RF capability and install a V572A transciever. did it 20 years ago and have had great rf coms ever since.

Yes I'm one of the few that have reliable RF  with my CM15A. However I rarely send any RF from my CM15A. HG by default sends PLC from it which is why for most people the CM15 is the better option over a CM19

I do believe @genemars new update to the drivers will no doubt improve things for you.

**Comment by tuicemen** [comment]
I'm wondering if we got off track from the OP, possibly due to the title.
The issue was the Web UI failed to send a command on first click of a device if no activity on the web page for some time.
This morning I came in started the PC then opened the HG Web UI Running on my Pi and clicked on the widget for my office light. Nothing happened and this Pi is connected to a Cm15.
a second press of the mouse turned on the light.
Usually  I double click on things in the computer so this may be why I never noticed this prior.
I don't believe this is related to drivers for either the cm15 or cm19 but more the Web UI as all my automation ran successfully during the night.

**Comment by alzzy** [comment]
Or there could be 2 issues!

**Comment by alzzy** [comment]
As I have HG running on win10 PC with stable.28 and with cm19a i don't have the double click issue.

**Comment by tuicemen** [comment]
I don't believe there is/was a driver issue as I stated my issues with the CM19 are range.
What I experienced this morning was what you first raised in your OP only difference being the controller.

**Comment by tuicemen** [comment]
> As I have HG running on win10 PC with stable.28 and with cm19a i don't have the double click issue.

So this proves it isn't a driver issue, nor a issue with HG, but the Pi Zero W.

**Comment by alzzy** [comment]
My issues with cm19a are not range but reliable triggering of cm19a. No led flash no device response regardless of how close i am to tranceiver. P;lease don't derail my issue. Open a new one.

**Comment by tuicemen** [comment]
I'm not derailing the issue as I experience this with the Cm19 As I have stated, and this is what your OP was about.
you stated:
>  it takes two clicks of an x10 module On or Off button in HG for the cm19a to transmit the command
 Its like the first click wakes it up and the second transmits the command.

**Comment by alzzy** [comment]
The driver recognized by win 10 and the one recognized by linux my be different or each OS may respond  differently or incorectly to the same code. Please open a new issue.

**Comment by alzzy** [comment]
In any case in the world of RF comms its always a good idea to send multiple commands.

**Comment by alzzy** [comment]
And by multiple I mean repetitive. That greatly increases the chance of the signal getting thru.

**Comment by genemars** [comment]
@tuicemen @alzzy **v1.2-stable.30** should fix a few minor issues with X10 and adds security modules DB to restore security modules on reboot/restart.
I published as stable, but CM15, CM19 and CM11 should be tested again.

**Comment by alzzy** [comment]
The above was with stable.29.  I will update to stable.30 and retest

**Comment by alzzy** [comment]
Updated to stable.30. Still experiencing unreliable performance to a single click on On or Off.

**Comment by alzzy** [comment]
I am Using Chrome. Don't know what Tuicemen used. Never used debugger so no,

**Comment by tuicemen** [comment]
updated to stable 30 using my phone last night and though unrelated it should be mentioned the installer from inside HG hung on the installing this didn't affect the install as a refresh opened HG with the new version installed. I don't recall having to do that when installing from a windows browser.

I've tested Stable 30 just with the CM15 though have it loaded on another pi for the cm19.
the Cm15 still requires a second button press to send a command (wake up) HG just for kicks I tested another PI HG that has Zwave switches and they don't require a second press to wake up HG.

Also I'm using the Windows default browser edge but I also experience this from my Android Phone browser.

**Comment by alzzy** [comment]
I also had the installer hang on both stable.29 and stable.30. Didn't do it on win10 install of the latest win release stable.26

**Comment by alzzy** [comment]
FYI, I reported that as issue #354 yesterday

**Comment by mralapete** [comment]
Guys when you experience a problem with HG you need to run it in debug mode so your actual problem can be identified. There are so many variables at play here like browser type, OS, hardware etc that can cause the issue and not necessarily HG. Do you know how to run it in debug mode and report the issue

@genemars I thought running HG in debug mode and reporting the issue this way was standard procedure. It must be getting very time consuming for you trying to figure out these recent problems without it particularly as you are currently working on a different project.

**Comment by alzzy** [comment]
Here is some log data.
[Module Double click log123118.txt](https://github.com/genielabs/HomeGenie/files/2718493/Module.Double.click.log123118.txt)

**Comment by alzzy** [comment]
I did a manual install of stable.30. Still have the double click issue. I did notice that when I single click and the cm19a **does not transmit,** the flag in the HG gui in the upper right corner comes out after 4 or 5 seconds after the click announcing the command but the cm19a does not respond.

**Comment by genemars** [comment]
I'm trying this on my laptop and it's working. Whenever I submit a command the CM19 led flashes and the command is performed with no visible delay.
So are you both @alzzy and @tuicemen experiencing the same problem?
The log @azzly sent looks fine... this is mine:
```
2019-01-01 13:46:53.0602 Info WebServiceGateway	x.x.x.x	HTTP	GET	200 /api/HomeAutomation.X10/C6/Control.On/undefined/ [OPEN]
2019-01-01 13:46:53.0602 Debug 20-40-BF-50-AF
2019-01-01 13:46:53.1943 Debug 20-40-BF-50-AF
2019-01-01 13:46:53.3305 Debug 20-40-BF-50-AF
2019-01-01 13:46:53.4665 Info HomeAutomation.X10	C6	X10 Module	Status.Level	1
2019-01-01 13:46:53.4785 Info WebServiceGateway	x.x.x.x	HTTP	GET	200 /api/HomeAutomation.X10/C6/Control.On/undefined/ [CLOSED AFTER 0,418 seconds]
2019-01-01 13:46:54.6947 Info WebServiceGateway	x.x.x.x	HTTP	GET	200 /api/HomeAutomation.X10/C6/Control.Off/undefined/ [OPEN]
2019-01-01 13:46:54.6953 Debug 20-40-BF-70-8F
2019-01-01 13:46:54.8263 Debug 20-40-BF-70-8F
2019-01-01 13:46:54.9625 Debug 20-40-BF-70-8F
2019-01-01 13:46:55.0987 Info HomeAutomation.X10	C6	X10 Module	Status.Level	0
```

If the CM19 led is not flashing then it means the command is not sent for some reason.
Have you tried the CM19 on a different system than RPi. Is it working? Is it a RPi-only related issue?
Have you tried different house-codes? I see you've been trying with "E" house code, do you have same result trying with a different house code (eg. "A")?

**Comment by alzzy** [comment]
I have HG installed on a win 10 pc and I **do not** have this issue with the cm19a. It is very reliable. I do **still see it** on the RPi. On Win 10 when I send a command the LED flashes and the command is sent and the device responds. I also agree "No flash, no command sent"

**Comment by alzzy** [comment]
PS, It does not appear to be house code dependent in my case

**Comment by alzzy** [comment]
Also it appears to be related somehow to the web gui as tuicemen reported he saw the issue with the **cm15** when triggered from the browser but that the overnight responses to HA bridge commands were successful. Also I know of another individual using the cm19a on a Pi Zero W with a different sd image and no web gui who gets reliable cm19a performance.

**Comment by alzzy** [comment]
I thought that this in my earlier post might be a clue " I did notice that when I single click and the cm19a does not transmit, the flag in the HG gui in the upper right corner comes out after 4 or 5 seconds after the click announcing the command but the cm19a does not respond."

**Comment by genemars** [comment]
@alzzy let's check **libusb** again from HG folder:
```
ls -la lib/mib

-rwxrw-r-- 1 gene gene   24064 dic 28 18:06 CM19Lib.dll
lrwxrwxrwx 1 root root      37 gen  1 13:42 libusb-1.0.so ->
-rwxrw-r-- 1 gene gene    6144 set  6 13:34 LibUsb.Common.dll
-rwxrw-r-- 1 gene gene  158208 set  6 13:35 LibUsbDotNet.LibUsbDotNet.dll
-rwxrw-r-- 1 gene gene   50688 dic 29 22:18 MIG.HomeAutomation.dll
-rwxrw-r-- 1 gene gene   31232 set 24 17:56 MIG.Protocols.dll
-rw-r--r-- 1 gene gene 2406991 set 25 16:03 p1db_custom.xml
```
is the **libusb-1.0.so** symbolic link pointing to the right location?
This might be an issue with with the **libusb**  path in which case also the CM15 driver should be broken.

**Comment by alzzy** [comment]
Tuicemen, Can you check this> i have no idea how!

**Comment by alzzy** [comment]
I am not sure that is a valid conclusion as the issue was also seen with the cm15a which **does not** rely on the Pi for power. Furthermore , Tuicemen noticed that the system with cm15a ran fine when triggered by the HA bridge but sometimes took a double click when triggered from the web gui. So I don't agree.

**Comment by genemars** [comment]
@alzzy the cm15 does not rely on usb for powering itself but it will still absorb some power when attached to the usb port.
What power supply are you using?
Using the web UI will involve network communication, that will require more power I guess.
I still believe this might be the cause.

**Comment by genemars** [comment]
@alzzy I'm now using a **5.2V 1.35A**  power supply and everything is running fine on a RPi Zero (v1.2 2015) with a 4 port USB switch, TP-link wifi dongle, CM19.

**Comment by alzzy** [comment]
I am confused. So what is the fix? You saying my 2.5 A 5v supply is not adequate?

**Comment by genemars** [comment]
@alzzy I'm confused too... it lasted for a few minutes then it hung and I had to reboot.
Does anyone tried upgrading mono from xamarin repos? I'm gonna try this too.

**Comment by alzzy** [comment]
genemars, Let me know what u learn?

**Comment by genemars** [comment]
This time after the reboot I did wait like 5 minutes before opening the web ui and its working. Not sure if it will last. 🙂

**Comment by alzzy** [comment]
Well that explanation doesn't seem to fit either as I know someone using the same hardware but with some what different software and he has excellent performance with cm19.

**Comment by alzzy** [comment]
Well I guess I am dead in the water then as they say. The implementation that works using the same hardware is triggered by a voice command via alexa and that appears reliable, in my situation it is triggered by a web gui and it isn't reliable. Does that provide any focus?

**Comment by mralapete** [comment]
@genemars in all this discussion have you considered that the addition of the HA Bridge server on the same Raspberry Pi Zero W on top of HG is creating an additional load on the low spec Pi processor and memory. The OP should either disable the HA Bridge and run his tests again or move his system to the more powerful Raspberry Pi 3+.

**Comment by genemars** [comment]
let's try this:
- ssh your pi
- cd /usr/local/bin/homegenie/html
- wget -O index2.html http://127.0.0.1/hg/html/index.html
- open the web ui using http://<rpi_address>/hg/html/index2.html

**Comment by alzzy** [comment]
on the wget is that -O the letter cap. O

**Comment by genemars** [comment]
@alzzy yes it is uppercase `-O`.

**Comment by alzzy** [comment]
i get a 404 error not found

**Comment by alzzy** [comment]
on the putty screen in response to the wget

**Comment by genemars** [comment]
@alzzy what port is hg running on?

**Comment by alzzy** [comment]
8080

**Comment by genemars** [comment]
so change the url in the wget command accordingly
```
- `wget -O index2.html http://127.0.0.1:8080/hg/html/index.html`
- open the web ui using `http://<rpi_address>:8080/hg/html/index2.html`
```

**Comment by alzzy** [comment]
No joy! same problem.

**Comment by alzzy** [comment]
Its like if you give it  enough time between toggling a device on or off, the Pi goes to sleep and  it takes 1 click to wake it up and a second click to send the command.

**Comment by alzzy** [comment]
and that time needs to be many seconds, sometimes even longer.

**Comment by genemars** [comment]
alzzly, tried disabling HA bridge?

**Comment by alzzy** [comment]
no, how do I do that

**Comment by alzzy** [comment]
oh in the gui. I can stop the bridge. I will try that!

**Comment by alzzy** [comment]
No doesn't help

**Comment by alzzy** [comment]
Do i need to undo the index2 somehow?

**Comment by genemars** [comment]
@alzzy not necessary.
This morning my RPi zero was still running happily with no issues.
So the only moment I've been able to reproduce this issue was in the beginning of my tests when I had problems with the USB HUB connector and the power supply.
In that case the problems with the USB I/O where already evident even before HomeGenie was installed, any extra device connected to the USB HUB was making the whole system unreliable and slow.

**Comment by alzzy** [comment]
Gene has put considerable effort into trying to id and correct this issue. I have helped with some of the testing. The results seem to confirm the conclusion that the PiX10 hub is not compatible with the cm19a. It seems to be a systems issue, ie HW + SW, as the cm19a does seem to play well with the Pi Zero w in one known instance where it is used as an Alexa Hub. Or that could be a "Random Success".

**Comment by alzzy** [comment]
The fact that one user has reported a similar issue with the cm15 is bothersome wrt the PiX10 Hub application.

**Comment by mralapete** [comment]
Gene just to confirm that the CM19 driver is working solid on both the Raspberry Pi Zero W and the Raspberry Pi 3+. HG in general is a lot more responsive on the Raspberry Pi 3+ probably because both the memory is double that of the Raspberry Pi Zero W and the processor is quite a bit faster and there's no need for USB powered hubs.

**Comment by genemars** [comment]
I just published **v1.2-stable.31** which now allow using UPnP library in automation programs.
So I created an example program that simulates a Belkin Switch so that Alexa can discover it.
Anyone who can test this with Alexa??
To test this:
- import the attacched automation program
- eventually modify the first line to point to a module of your pick
- Action->Run to start the program that will create virtual UPnP Belkin Switch that should be also discovered by Alexa
If we make this work maybe HA bridge will not be required anymore? Right?

[1000-UPnP_Test_for_Alexa.hgx.txt](https://github.com/genielabs/HomeGenie/files/2725261/1000-UPnP_Test_for_Alexa.hgx.txt)

**Comment by tuicemen** [comment]
@genemars
 I been tring to fight off a bad christmas virus that got ito my eyes so ive not been able to comment on things as late I'll attemp to test this as soon as I can see straight.

**Comment by mralapete** [comment]
Initial test of your Belkin Switch program. Alexa not discovering a chosen device

**Comment by mralapete** [comment]
@genemars could you open this as a separate issue
